test-framework/common/src/main/java/io/quarkus/test/common/PathTestHelper.java

```diff
@@ -45,8 +62,23 @@ public static Path getTestClassesLocation(Class<?> testClass) {
         return path.getRoot().resolve(path.subpath(0, path.getNameCount() - Paths.get(classFileName).getNameCount()));
     }
 
+    /**
+     * Resolves the directory or the JAR file containing the application being tested by the test class.
+     *
+     * @param testClass the test class
+     * @return directory or JAR containing the application being tested by the test class
+     */
     public static Path getAppClassLocation(Class<?> testClass) {
-        String testClassPath = getTestClassesLocation(testClass).toString();
+        final String testClassPath = getTestClassesLocation(testClass).toString();
+        if (testClassPath.endsWith(".jar")) {
+            if (testClassPath.endsWith("-tests.jar")) {
```

Comment:
I think at some point we need to revisit this and provide some way for the execution environment to pass in this information, so it is not hard coded like this. This is fine for the moment though.